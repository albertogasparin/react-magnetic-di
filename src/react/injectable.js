import { diRegistry } from './constants';
import { getDisplayName, warnOnce } from './utils';

export function injectable(
  from,
  implementation,
  { displayName, target, track = true, global = false } = {}
) {
  let impl = implementation;
  if (typeof impl === 'function') {
    impl.displayName =
      displayName || getDisplayName(impl) || getDisplayName(from, 'di');
  } else if (typeof impl !== 'object') {
    impl = {
      [Symbol.toPrimitive]() {
        return implementation;
      },
    };
  }

  if (diRegistry.has(impl) && diRegistry.get(impl).from !== from) {
    warnOnce(
      `You are trying to use replacement "${
        displayName || impl.displayName
      }" on multiple injectables. ` +
        `That will override only the last dependency, as each replacement is uniquely linked.`
    );
  }
  diRegistry.set(impl, {
    value: implementation,
    from,
    targets: target && (Array.isArray(target) ? target : [target]),
    track,
    global,
    cause: track
      ? new Error(
          'Injectable created but not used. If this is on purpose, add "{track: false}"'
        )
      : null,
  });
  return impl;
}
