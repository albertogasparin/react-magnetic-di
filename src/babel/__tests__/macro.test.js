/* eslint-env jest */
import { transform } from '@babel/core';
import plugin from 'babel-plugin-macros';

const babel = (code, { options, env } = {}) =>
  transform(code, {
    filename: 'noop.js',
    presets: [['@babel/preset-react', { development: false, pragma: '__jsx' }]],
    plugins: [[plugin, options]],
    babelrc: false,
    configFile: false,
    sourceType: 'module',
    caller: { name: 'tests', supportsStaticESM: true },
    envName: env,
  }).code;

describe('macro plugin', () => {
  it('should work in class components', () => {
    const input = `
        import React, { Component } from 'react';
        import di from './src/babel/macro';
        import Modal from 'modal';
  
        class MyComponent extends Component {
          render() {
            di(Modal);
            return <Modal />;
          }
        }
      `;
    expect(babel(input)).toMatchSnapshot();
  });

  it('should work with renamed default and functional components', () => {
    const input = `
        import React from 'react';
        import injectable from './src/babel/macro';
        import Modal from 'modal';

        const MyComponent = () => {
          injectable(Modal);
          return <Modal />;
        };
      `;
    expect(babel(input)).toMatchSnapshot();
  });
});
