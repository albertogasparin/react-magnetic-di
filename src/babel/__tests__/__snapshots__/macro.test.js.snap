// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro plugin should work in class components 1`] = `
"import { di } from \\"react-magnetic-di\\";
import React, { Component } from 'react';
import Modal from 'modal';

class MyComponent extends Component {
  render() {
    var [_Modal] = di([Modal], MyComponent);
    return __jsx(_Modal, null);
  }

}"
`;

exports[`macro plugin should work with renamed default and functional components 1`] = `
"import { di as injectable } from \\"react-magnetic-di\\";
import React from 'react';
import Modal from 'modal';

const MyComponent = () => {
  var [_Modal] = injectable([Modal], MyComponent);
  return __jsx(_Modal, null);
};"
`;
